### PR TITLE
[EWS] Support PRs when sending layout test failure emails

### PR DESCRIPTION
**Pull request by JonWBedard**
#### ed081365ebf7e362bf61a93d527a14a4319e48c6
<pre>
[EWS] Support PRs when sending layout test failure emails
<a href="https://bugs.webkit.org/show_bug.cgi?id=236654">https://bugs.webkit.org/show_bug.cgi?id=236654</a>
&lt;rdar://problem/88974170 &gt;

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(AnalyzeLayoutTestsResults.send_email_for_new_test_failures): Draft different
email for failed patch than failed pull request build.

Canonical link: <a href="https://commits.webkit.org/247326@main">https://commits.webkit.org/247326@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@289888">https://svn.webkit.org/repository/webkit/trunk@289888</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Tools/Scripts/webkitpy/layout_tests/servers/websocket_server.py [hunk 1]
```diff
@@ -116,11 +116,11 @@ def _prepare_config(self):
 
         # FIXME https://bugs.webkit.org/show_bug.cgi?id=206546: Should be using the same version of Python run-webkit-tests is
         python_interp = sys.executable
-        if sys.version_info > (3, 0):
-            python_interp = 'python2'
+        if sys.version_info < (3, 0):
```
Comment (aj062) [inline Tools/Scripts/webkitpy/layout_tests/servers/websocket_server.py]: Seems like this unrelated change got included by mistake

File: Tools/CISupport/ews-build/steps.py [hunk 2]
```diff
@@ -3026,11 +3026,36 @@ def send_email_for_pre_existing_failure(self, test_name):
...
+                change_string = 'Patch {}'.format(patch_id)
+            elif pr_number and sha:
+                change_string = 'Hash {}'.format(sha)
+                change_author, errors = GitHub.email_for_owners(self.getProperty('owners', []))
+                for error in errors:
+                    print(error)
+                    self._addToLog('stdio', error)
+
+            if not change_author:
+                self._addToLog('Failed to determine change author for hash {} belonging to PR #{}'.format(sha, patch_id))
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Same comment here as https://github.com/WebKit/WebKit/pull/144#discussion_r807377095

**Review by aj062 — APPROVED** [review]

File: Tools/Scripts/webkitpy/layout_tests/servers/websocket_server.py (hunk 1 above)
Comment (JonWBedard) [reply Tools/Scripts/webkitpy/layout_tests/servers/websocket_server.py]: Not an unrelated change, a deliberate breakage I'm sending through the UAT instance to test this. Will not be landing this bit.

Comment (JonWBedard) [reply Tools/Scripts/webkitpy/layout_tests/servers/websocket_server.py]: https://ews-build.webkit-uat.org/#/builders/81/builds/26 should trigger this email

**Comment by JonWBedard** [comment]
Example of an email generated with this code:
```
EWS has detected layout test failures on macOS-BigSur-Release-WK1-Tests-EWS while testing [Hash 0cef9e68](https://github.com/WebKit/WebKit/commit/0cef9e68) for [PR #147](https://github.com/WebKit/WebKit/pull/147).

Full details are available at: https://ews-build.webkit-uat.org/#/builders/81/builds/32

Change author: [jbedard@apple.com](mailto:jbedard@apple.com)

Layout test failures:

- css3/calc/transitions-dependent.html ([test history](https://results.webkit.org/?suite=layout-tests&test=css3/calc/transitions-dependent.html))
- http/tests/websocket/tests/hybi/client-close.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/client-close.html))
- http/tests/websocket/tests/hybi/close-code-and-reason.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/close-code-and-reason.html))
- http/tests/websocket/tests/hybi/deflate -extension.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/deflate-extension.html))
- http/tests/websocket/tests/hybi/deflate-frame-comp-bit-onoff.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/deflate-frame-comp-bit-onoff.html))
- http/tests/websocket/tests/hybi/deflate-frame-set-bfinal.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/deflate-frame-set-bfinal.html))
- http/tests/websocket/tests/hybi/extensions.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/extensions.html))
- http/tests/websocket/tests/hybi/fragmented-binary-frames.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/fragmented-binary-frames.html))
- http/tests/websocket /tests/hybi/frame-lengths.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/frame-lengths.html))
- http/tests/websocket/tests/hybi/handshake-ok-with-http-version-beyond-1_1.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/handshake-ok-with-http-version-beyond-1_1.html))
- http/tests/websocket/tests/hybi/handshake-ok-with-legacy-websocket-response-headers.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/handshake-ok-with-legacy-websocket-response-headers.html))
- http/tests/websocket/tests/hybi/httponly-cookie.pl ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/httponly-cookie.pl))
- http/tests/websocket/tests/hybi/inspector/binary.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websoc%20ket/tests/hybi/inspector/binary.html))
- http/tests/websocket/tests/hybi/inspector/send-and-receive.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/inspector/send-and-receive.html))
- http/tests/websocket/tests/hybi/multiple-subprotocols.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/multiple-subprotocols.html))
- http/tests/websocket/tests/hybi/null-character.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/null-character.html))
- http/tests/websocket/tests/hybi/pong.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/pong.html))
- http/tests/websocket/tests/hybi/receive-arraybuffer.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/receive%20-arraybuffer.html))
- http/tests/websocket/tests/hybi/receive-blob.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/receive-blob.html))
- http/tests/websocket/tests/hybi/reserved-bits.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/reserved-bits.html))
- http/tests/websocket/tests/hybi/secure-cookie-insecure-connection.pl ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/secure-cookie-insecure-connection.pl))
- http/tests/websocket/tests/hybi/secure-cookie-secure-connection.pl ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/secure-cookie-secure-connection.pl))
- http/tests/websocket/tests/hybi/workers/close-code-and-reason.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tes%20ts/websocket/tests/hybi/workers/close-code-and-reason.html))
- http/tests/websocket/tests/hybi/workers/multiple-subprotocols.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/multiple-subprotocols.html))
- http/tests/websocket/tests/hybi/workers/no-onmessage-in-sync-op.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/no-onmessage-in-sync-op.html))
- http/tests/websocket/tests/hybi/workers/receive-arraybuffer.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/receive-arraybuffer.html))
- http/tests/websocket/tests/hybi/workers/receive-blob.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/receive-blob.html))
- http/tests/websocket/tests/hybi/workers/send-arraybu ffer.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/send-arraybuffer.html))
- http/tests/websocket/tests/hybi/workers/send-arraybufferview.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/send-arraybufferview.html))
- http/tests/websocket/tests/hybi/workers/send-blob.html ([test history](https://results.webkit.org/?suite=layout-tests&test=http/tests/websocket/tests/hybi/workers/send-blob.html))

To unsubscribe from these notifications or to provide any feedback please email [aakash_jain@apple.com](mailto:aakash_jain@apple.com)
```

**Comment by JonWBedard** [comment]
Landed https://commits.webkit.org/247326@main (ed081365ebf7e362bf61a93d527a14a4319e48c6)!
